src/java/org/apache/cassandra/tcm/RemoteProcessor.java

```diff
@@ -201,51 +203,45 @@ public static <REQ, RSP> RSP sendWithCallback(Verb verb, REQ request, CandidateI
 
     public static <REQ, RSP> void sendWithCallbackAsync(Promise<RSP> promise, Verb verb, REQ request, CandidateIterator candidates, Retry retryPolicy)
     {
-        class Request implements RequestCallbackWithFailure<RSP>
-        {
-            void retry()
-            {
-                if (promise.isCancelled() || promise.isDone())
-                    return;
-                if (Thread.currentThread().isInterrupted())
-                    promise.setFailure(new InterruptedException());
-                if (!candidates.hasNext())
-                    promise.tryFailure(new IllegalStateException(String.format("Ran out of candidates while sending %s: %s", verb, candidates)));
-
-                MessagingService.instance().sendWithCallback(Message.out(verb, request), candidates.next(), this);
-            }
-
-            @Override
-            public void onResponse(Message<RSP> msg)
-            {
-                promise.trySuccess(msg.payload);
-            }
-
-            @Override
-            public void onFailure(InetAddressAndPort from, RequestFailureReason reason)
-            {
-                if (reason == RequestFailureReason.NOT_CMS)
-                {
-                    logger.debug("{} is not a member of the CMS, querying it to discover current membership", from);
-                    DiscoveredNodes cms = tryDiscover(from);
-                    candidates.addCandidates(cms);
-                    candidates.timeout(from);
-                    logger.debug("Got CMS from {}: {}, retrying on: {}", from, cms, candidates);
-                }
-                else
-                {
-                    candidates.timeout(from);
-                    logger.warn("Got error from {}: {} when sending {}, retrying on {}", from, reason, verb, candidates);
-                }
-
-                if (retryPolicy.reachedMax())
-                    promise.tryFailure(new IllegalStateException(String.format("Could not succeed sending %s to %s after %d tries", verb, candidates, retryPolicy.tries)));
-                else
-                    retry();
-            }
-        }
-
-        new Request().retry();
+        //TODO (now): the retry defines how long to wait for a retry, but the old behavior scheduled the message right away... should this be delayed as well?
+        MessagingService.instance().<REQ, RSP>sendWithRetries(Backoff.fromRetry(retryPolicy), MessageDelivery.ImmediateRetryScheduler.instance,
+                                                              verb, request, candidates,
+                                                              (attempt, success, failure) -> {
```

Comment:
nit: maybe rename `failure` to throwable?

Reply:
I prefer `failure` in this case as it matches the `promise`

```
tryFailure(failure)
```